### ISSUE TITLE
v100 GPU vs CPU different accuracy，the same dataset and parameter

### ISSUE THREAD

**Issue by DenceChen** [dmlc/xgboost#4978]
I am training the same dataset on GPU and CPU with xgboost, but got different accuracy on  eval data.
The same data with the same parameters using the xgboost GPU version and the CPU version training model are not the same, the test set is the same.

**gpu:**
model = xgb.XGBRegressor(n_estimators=800,
                         objective = 'reg:linear',
                         booster='gbtree',
                         learning_rate = 0.05,
                         gamma=2,
                         max_depth=9,
                         min_child_weight=5,
                         subsample=0.9,
                         colsample_bytree=0.9, 
                         reg_lambda=5,
                         reg_alpha=3,
                         updater="grow_gpu",
                         seed=7,
                         n_jobs=-1)

model.fit(X_train, y_train, early_stopping_rounds=20, eval_metric="rmse", eval_set=[(X_train, y_train), (X_test, y_test)])

![image](https://user-images.githubusercontent.com/11643704/67389393-21761b80-f5cd-11e9-998c-30c1e265f06f.png)

*********************************************************************************************************

**cup:**
model = xgb.XGBRegressor(n_estimators=800,
                         objective = 'reg:linear',
                         booster='gbtree',
                         learning_rate = 0.05,
                         gamma=2,
                         max_depth=9,
                         min_child_weight=5,
                         subsample=0.9,
                         colsample_bytree=0.9, 
                         reg_lambda=5,
                         reg_alpha=3,
                         n_jobs=-1)

model.fit(X_train, y_train, early_stopping_rounds=20, eval_metric="rmse", eval_set=[(X_train, y_train), (X_test, y_test)])

![image](https://user-images.githubusercontent.com/11643704/67389466-44083480-f5cd-11e9-92df-cd84ff179bf0.png)

**Comment by trivialfis** [comment]
On CPU, use `tree_method = hist`, on GPU, use `tree_method = gpu_hist`.  These two are same algorithms.  There's still minor discrepancy due to floating point error.

**Comment by trivialfis** [comment]
And don't set the updater.  There is a warning I think.

**Comment by DenceChen** [comment]
@trivialfis thanks for your reply, I tried your suggestion but the effect is still not working.

**Comment by DenceChen** [comment]
@trivialfis the CPU result better than GPU

**Comment by trivialfis** [comment]
Another thing is, the gamma parameter is only supported on gpu in a recent commit as a pre-pruning step.  It's used for preventing over fitting.  We are trying to make a new release, it might take sometime as this one is quite big, so you may want to build it from source for that parameter.
